pynumaflow/reducestreamer/servicer/task_manager.py

```diff
@@ -0,0 +1,258 @@
+            curr_task = ReduceResult(task, niter, keys, req.windows[0], res_queue, consumer)
+
+            # Save the result of the reduce operation to the task list
+            self.tasks[unified_key] = curr_task
+
+        # Put the request in the iterator
+        await curr_task.iterator.put(d)
+
+    async def send_datum_to_task(self, req):
+        """
+        Appends the request to the existing window reduce task.
+        If the task does not exist, create it.
+        """
+        if len(req.windows) != 1:
+            raise UDFError("reduce append operation error: invalid number of windows")
+        d = req.payload
+        keys = d.keys()
+        unified_key = build_unique_key_name(keys, req.windows[0])
+        result = self.tasks.get(unified_key, None)
+        if not result:
+            await self.create_task(req)
+        else:
+            # print(result)
+            await result.iterator.put(d)
+
+    async def __invoke_reduce(
+        self,
+        keys: list[str],
+        request_iterator: AsyncIterable[Datum],
+        output: NonBlockingIterator,
+        window: ReduceWindow,
+    ):
+        """
+        Invokes the UDF reduce handler with the given keys,
+        request iterator, and window. Returns the result of the
+        reduce operation.
+        """
+        new_instance = self.__reduce_handler
+
+        # Convert the window to a datetime object
+        start_dt = datetime.fromtimestamp(int(window.start.ToMilliseconds()) / 1e3, timezone.utc)
+        end_dt = datetime.fromtimestamp(int(window.end.ToMilliseconds()) / 1e3, timezone.utc)
+        interval_window = IntervalWindow(start_dt, end_dt)
+        md = Metadata(interval_window=interval_window)
+        # If the reduce handler is a class instance, create a new instance of it.
+        # It is required for a new key to be processed by a
+        # new instance of the reducer for a given window
+        # Otherwise the function handler can be called directly
+        if isinstance(self.__reduce_handler, _ReduceStreamBuilderClass):
+            new_instance = self.__reduce_handler.create()
+        try:
+            _ = await new_instance(keys, request_iterator, output, md)
+        # If there is an error in the reduce operation, log and
+        # then send the error to the result queue
+        except Exception as err:
+            _LOGGER.critical("UDFError, re-raising the error", exc_info=True)
+            # Put the exception in the result queue
+            await self.global_result_queue.put(err)
+
+    async def process_input_stream(self, request_iterator: AsyncIterable[reduce_pb2.ReduceRequest]):
+        # Start iterating through the request iterator and create tasks
+        # based on the operation type received.
+        try:
+            async for request in request_iterator:
+                # check whether the request is an open or append operation
+                if request.operation is int(WindowOperation.OPEN):
```

Comment:
Do we need to cast here? Have we checked that it doesn't work otherwise?

Reply:
Yeah! Without cast it was not working as expected